File: backend/app/schemas/profiles.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class CompanyProfileRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: uuid.UUID
    company_name: str
    description: str | None
    city: str | None
    regions_text: str | None
    project_types: str | None
    years_on_market: int | None
    contact_person: str | None
    phone: str | None
    messengers_text: str | None
    email_public: str | None
    cooperation_terms: str | None
    avg_budget_note: str | None
    payment_methods_text: str | None
    media_note: str | None
    avatar_url: str | None
    rating_avg: float
    reviews_count: int
    created_at: datetime
    updated_at: datetime


class CompanyProfileUpdate(BaseModel):
    company_name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    city: str | None = None
    regions_text: str | None = None
    project_types: str | None = None
    years_on_market: int | None = Field(default=None, ge=0, le=200)
    contact_person: str | None = None
    phone: str | None = None
    messengers_text: str | None = None
    email_public: str | None = None
    cooperation_terms: str | None = None
    avg_budget_note: str | None = None
    payment_methods_text: str | None = None
    media_note: str | None = None
    avatar_url: str | None = None


class WorkerProfileRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: uuid.UUID
    full_name: str
    profession: str | None
    specialization: str | None
    experience_years: int | None
    skills_text: str | None
    city: str | None
    willing_to_travel: bool
    desired_rate_note: str | None
    work_format: str | None
    bio: str | None
    documents_note: str | None
    portfolio_note: str | None
    avatar_url: str | None
    rating_avg: float
    reviews_count: int
    created_at: datetime
    updated_at: datetime


class WorkerProfileUpdate(BaseModel):
    full_name: str | None = Field(default=None, min_length=1, max_length=255)
    profession: str | None = None
    specialization: str | None = None
    experience_years: int | None = Field(default=None, ge=0, le=80)
    skills_text: str | None = None
    city: str | None = None
    willing_to_travel: bool | None = None
    desired_rate_note: str | None = None
    work_format: str | None = None
    bio: str | None = None
    documents_note: str | None = None
    portfolio_note: str | None = None
    avatar_url: str | None = None


class BrigadeProfileRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: uuid.UUID
    name: str
    leader_name: str | None
    headcount: int
    roles_composition_text: str | None
    specialization: str | None
    past_objects_note: str | None
    regions_text: str | None
    has_tools: bool
    has_transport: bool
    avg_price_note: str | None
    availability_note: str | None
    bio: str | None
    portfolio_note: str | None
    avatar_url: str | None
    rating_avg: float
    reviews_count: int
    created_at: datetime
    updated_at: datetime


class BrigadeProfileUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    leader_name: str | None = None
    headcount: int | None = Field(default=None, ge=1, le=500)
    roles_composition_text: str | None = None
    specialization: str | None = None
    past_objects_note: str | None = None
    regions_text: str | None = None
    has_tools: bool | None = None
    has_transport: bool | None = None
    avg_price_note: str | None = None
    availability_note: str | None = None
    bio: str | None = None
    portfolio_note: str | None = None
    avatar_url: str | None = None

File: frontend/src/types/profiles.ts
/** Согласовано с backend/app/schemas/profiles.py (Read) */

export type CompanyProfile = {
  user_id: string;
  company_name: string;
  description: string | null;
  city: string | null;
  regions_text: string | null;
  project_types: string | null;
  years_on_market: number | null;
  contact_person: string | null;
  phone: string | null;
  messengers_text: string | null;
  email_public: string | null;
  cooperation_terms: string | null;
  avg_budget_note: string | null;
  payment_methods_text: string | null;
  media_note: string | null;
  avatar_url: string | null;
  rating_avg: number;
  reviews_count: number;
  created_at: string;
  updated_at: string;
};

export type WorkerProfile = {
  user_id: string;
  full_name: string;
  profession: string | null;
  specialization: string | null;
  experience_years: number | null;
  skills_text: string | null;
  city: string | null;
  willing_to_travel: boolean;
  desired_rate_note: string | null;
  work_format: string | null;
  bio: string | null;
  documents_note: string | null;
  portfolio_note: string | null;
  avatar_url: string | null;
  rating_avg: number;
  reviews_count: number;
  created_at: string;
  updated_at: string;
};

export type BrigadeProfile = {
  user_id: string;
  name: string;
  leader_name: string | null;
  headcount: number;
  roles_composition_text: string | null;
  specialization: string | null;
  past_objects_note: string | null;
  regions_text: string | null;
  has_tools: boolean;
  has_transport: boolean;
  avg_price_note: string | null;
  availability_note: string | null;
  bio: string | null;
  portfolio_note: string | null;
  avatar_url: string | null;
  rating_avg: number;
  reviews_count: number;
  created_at: string;
  updated_at: string;
};

function str(v: unknown): string {
  return typeof v === "string" ? v : "";
}

function num(v: unknown): number | null {
  if (v === null || v === undefined) return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

function bool(v: unknown, fallback: boolean): boolean {
  return typeof v === "boolean" ? v : fallback;
}

export function parseCompanyProfile(raw: Record<string, unknown>): CompanyProfile {
  return {
    user_id: str(raw.user_id),
    company_name: str(raw.company_name) || "Компания",
    description: raw.description != null ? String(raw.description) : null,
    city: raw.city != null ? String(raw.city) : null,
    regions_text: raw.regions_text != null ? String(raw.regions_text) : null,
    project_types: raw.project_types != null ? String(raw.project_types) : null,
    years_on_market: num(raw.years_on_market),
    contact_person: raw.contact_person != null ? String(raw.contact_person) : null,
    phone: raw.phone != null ? String(raw.phone) : null,
    messengers_text: raw.messengers_text != null ? String(raw.messengers_text) : null,
    email_public: raw.email_public != null ? String(raw.email_public) : null,
    cooperation_terms: raw.cooperation_terms != null ? String(raw.cooperation_terms) : null,
    avg_budget_note: raw.avg_budget_note != null ? String(raw.avg_budget_note) : null,
    payment_methods_text: raw.payment_methods_text != null ? String(raw.payment_methods_text) : null,
    media_note: raw.media_note != null ? String(raw.media_note) : null,
    avatar_url: raw.avatar_url != null && String(raw.avatar_url).trim() ? String(raw.avatar_url) : null,
    rating_avg: Number(raw.rating_avg) || 0,
    reviews_count: Number(raw.reviews_count) || 0,
    created_at: str(raw.created_at),
    updated_at: str(raw.updated_at),
  };
}

export function parseWorkerProfile(raw: Record<string, unknown>): WorkerProfile {
  return {
    user_id: str(raw.user_id),
    full_name: str(raw.full_name) || "Специалист",
    profession: raw.profession != null ? String(raw.profession) : null,
    specialization: raw.specialization != null ? String(raw.specialization) : null,
    experience_years: num(raw.experience_years),
    skills_text: raw.skills_text != null ? String(raw.skills_text) : null,
    city: raw.city != null ? String(raw.city) : null,
    willing_to_travel: bool(raw.willing_to_travel, true),
    desired_rate_note: raw.desired_rate_note != null ? String(raw.desired_rate_note) : null,
    work_format: raw.work_format != null ? String(raw.work_format) : null,
    bio: raw.bio != null ? String(raw.bio) : null,
    documents_note: raw.documents_note != null ? String(raw.documents_note) : null,
    portfolio_note: raw.portfolio_note != null ? String(raw.portfolio_note) : null,
    avatar_url: raw.avatar_url != null && String(raw.avatar_url).trim() ? String(raw.avatar_url) : null,
    rating_avg: Number(raw.rating_avg) || 0,
    reviews_count: Number(raw.reviews_count) || 0,
    created_at: str(raw.created_at),
    updated_at: str(raw.updated_at),
  };
}

export function parseBrigadeProfile(raw: Record<string, unknown>): BrigadeProfile {
  return {
    user_id: str(raw.user_id),
    name: str(raw.name) || "Бригада",
    leader_name: raw.leader_name != null ? String(raw.leader_name) : null,
    headcount: Math.max(1, Number(raw.headcount) || 1),
    roles_composition_text: raw.roles_composition_text != null ? String(raw.roles_composition_text) : null,
    specialization: raw.specialization != null ? String(raw.specialization) : null,
    past_objects_note: raw.past_objects_note != null ? String(raw.past_objects_note) : null,
    regions_text: raw.regions_text != null ? String(raw.regions_text) : null,
    has_tools: bool(raw.has_tools, false),
    has_transport: bool(raw.has_transport, false),
    avg_price_note: raw.avg_price_note != null ? String(raw.avg_price_note) : null,
    availability_note: raw.availability_note != null ? String(raw.availability_note) : null,
    bio: raw.bio != null ? String(raw.bio) : null,
    portfolio_note: raw.portfolio_note != null ? String(raw.portfolio_note) : null,
    avatar_url: raw.avatar_url != null && String(raw.avatar_url).trim() ? String(raw.avatar_url) : null,
    rating_avg: Number(raw.rating_avg) || 0,
    reviews_count: Number(raw.reviews_count) || 0,
    created_at: str(raw.created_at),
    updated_at: str(raw.updated_at),
  };
}
